source/agora/consensus/data/UTXOSet.d

```diff
@@ -80,23 +151,92 @@ public class UTXOSet
 
         Params:
             tx = the transaction
+            height = the height of block
 
     ***************************************************************************/
 
-    public void updateUTXOCache (const ref Transaction tx) nothrow @safe
+    public void updateUTXOCache (const ref Transaction tx, ulong height) nothrow @safe
     {
-        foreach (const ref input; tx.inputs)
+        if (tx.type == TxType.Freeze)
+        {
+            if ((tx.inputs.length == 1) && (tx.outputs.length == 1))
+            {
+                auto input = tx.inputs[0];
+                auto utxo_hash = getHash(input.previous, input.index);
+                UTXOSetValue prev_utxo;
+
+                if (this.findUTXOSetValue (input.previous, input.index, prev_utxo))
```

Comment:
This seems to look for a previous frozen UTXO when a new frozen UTXO is added to the cache. So how do you add the very first frozen UTXO? It won't find a previous one, because a previous one doesn't exist yet.

The pull request passes the test-suite, but there are no tests for any of this code.

Reply:
Ah I was wrong, I see it in the code below. But, could you add tests?